packages/plugin/webpack/src/WebpackConfig.ts

```diff
@@ -163,114 +180,122 @@ export default class WebpackConfigGenerator {
+      electronPreload: [] as (WebpackPluginEntryPointPreloadOnly | WebpackPluginEntryPointLocalWindow)[],
+      sandboxedPreload: [] as (WebpackPluginEntryPointPreloadOnly | WebpackPluginEntryPointLocalWindow)[],
+    };
+    for (const entry of entryPoints) {
+      if (entry.nodeIntegration ?? this.pluginConfig.renderer.nodeIntegration) {
+        if (isPreloadOnly(entry)) {
+          entryPointsForTarget.electronPreload.push(entry);
+        } else {
+          entryPointsForTarget.electronRenderer.push(entry);
+          if (isLocalWindow(entry) && entry.preload) {
+            entryPointsForTarget.electronPreload.push(entry);
+          }
+        }
+      } else {
+        if (isPreloadOnly(entry)) {
+          entryPointsForTarget.sandboxedPreload.push(entry);
+        } else {
+          entryPointsForTarget.web.push(entry);
+          if (isLocalWindow(entry) && entry.preload) {
+            entryPointsForTarget.sandboxedPreload.push(entry);
+          }
+        }
+      }
     }
+    const rendererConfigPromises = [
+      this.buildRendererConfig(entryPointsForTarget.web, RendererTarget.Web),
+      this.buildRendererConfig(entryPointsForTarget.electronRenderer, RendererTarget.ElectronRenderer),
+      this.buildRendererConfig(entryPointsForTarget.electronPreload, RendererTarget.ElectronPreload),
+      this.buildRendererConfig(entryPointsForTarget.sandboxedPreload, RendererTarget.SandboxedPreload),
+    ];
+
+    const rendererConfig = await Promise.all(rendererConfigPromises).then((res) => {
+      return res.filter(function isNotNull<T>(item: T | null): item is T {
+        return !!item;
+      });
+    });
+    return rendererConfig;
+  }
 
-    const rendererConfig = await this.resolveConfig(entryPoint.preload.config || this.pluginConfig.renderer.config);
-    const prefixedEntries = entryPoint.prefixedEntries || [];
-
-    return webpackMerge(
-      {
-        devtool: this.rendererSourceMapOption,
-        mode: this.mode,
-        entry: prefixedEntries.concat([entryPoint.preload.js]),
-        output: {
-          path: path.resolve(this.webpackDir, 'renderer', entryPoint.name),
-          filename: 'preload.js',
-        },
-        node: {
-          __dirname: false,
-          __filename: false,
-        },
+  buildRendererBaseConfig(target: RendererTarget): webpack.Configuration {
+    return {
+      target: rendererTargetToWebpackTarget(target),
+      devtool: this.rendererSourceMapOption,
+      mode: this.mode,
+      output: {
+        path: path.resolve(this.webpackDir, 'renderer'),
+        filename: '[name]/index.js',
+        globalObject: 'self',
+        ...(this.isProd ? {} : { publicPath: '/' }),
       },
-      rendererConfig || {},
-      { target: 'electron-preload' }
-    );
+      node: {
+        __dirname: false,
+        __filename: false,
+      },
+      plugins: [new AssetRelocatorPatch(this.isProd, !!this.pluginConfig.renderer.nodeIntegration)],
```

Comment:
I know this code existed before but it's semantically incorrect

```suggestion
      plugins: [new AssetRelocatorPatch(this.isProd, target === RendererTarget.ElectronRenderer || target === RendererTarget.ElectronPreload)],
```